### DISCUSSION TITLE
Is full duplex mode available?

### DISCUSSION THREAD

**Discussion by AskarZinurov** [socketry/async-websocket#36] [General]
Hi, @ioquatix! First of all I want to thank you for such awesome job on async ruby gems here. I hope much of your work will go to ruby std lib 🤞
I am playing with Falcon and async-websockets and trying to implement some kind of pub/sub functionality. I was faced with problem what websocket connection becomes blocked by `message = connection.read` code. I mean what I cannot write to it or set a read timeout to break fiber from stuck. This happens because websocket behaves like usual socket and can be used only one way at a time?

**Comment by ioquatix** [comment]
Please check out the getting started guide: https://socketry.github.io/async-websocket/guides/getting-started/index.html

**Reply by AskarZinurov** [reply]
Thank you, the only missing part was `connection.flush`. Works like a charm!

**Reply by ioquatix** [reply]
Buffering can make or break performance, if you are sending 100 messages, send them all and call flush at the end. The underlying protocol can't know the most optimal flushing strategy - I've tried to make this implicit with timers in the past but it was a poor substitute for application level flushing.
